Added YOLOF object detection model (#2561)
openvinotoolkit/open_model_zoo: models/public/yolof/README.md

| Metric                                                                | Value  |
| --------------------------------------------------------------------- | -------|
| mAP                                                                   | 60.41% |
| [COCO mAP (0.5)](https://cocodataset.org/#detection-eval)             | 66.14% |
| [COCO mAP (0.5:0.05:0.95)](https://cocodataset.org/#detection-eval)   | 43.51% |

## Input

### Original model

Image, name - `image_input`, shape - `1, 3, 608, 608`, format is `B, C, H, W`, where:

- `B` - batch size
- `C` - channel
- `H` - height
- `W` - width

Channel order is `BGR`.
Mean values - [103.53, 116.28, 123.675].

### Converted model

Image, name - `image_input`, shape - `1, 3, 608, 608`, format is `B, C, H, W`, where:

- `B` - batch size
- `C` - channel
- `H` - height
- `W` - width

Channel order is `BGR`.

## Output

### Original model

The detected boxes, scores and labels for each box. The postprocessing is implemented inside the model and is performed while inference of model.

Detection box has format [`x1`, `y1`, `x2`, `y2`], where:

- (`x1`, `y1`) - left-bottom corner

> - (`x2`, `y2`) - right-top corner

Reviewer: Update description for original output

### Converted model

The array of detection summary info, name - `boxes`, shape - `1, 8664, 84`. The anchor values are `16,16,  32,32,  64,64,  128,128,  256,256,  512,512`.

For each case format is `B, N*Cx*Cy, 84`, where

- `B` - batch size
- `N` - number of detection boxes for cell (N = 6)
- `Cx`, `Cy` - cell index

Detection box has format [`x`, `y`, `h`, `w`, `class_no_1`, ..., `class_no_80`], where:

- (`x`, `y`) - coordinates of box center relative to the cell
- `h`, `w` - raw height and width of box, apply [exponential function](https://en.wikipedia.org/wiki/Exponential_function) and multiply by corresponding anchors to get absolute height and width values

> - `class_no_1`,...,`class_no_80` - probability distribution over the classes in the [0,1] range, multiply by confidence value to get confidence of each class

Reviewer: Fix the description, it differs from the wrapper's code

# yolof

## Use Case and High-Level Description

> YOLOF is a simple, fast, and efficient object detector without FPN. Model based on ["You Only Look One-level Feature"](https://arxiv.org/abs/2103.09460) paper. It was implemented in PyTorch\* framework. For details see [repository](https://github.com/megvii-model/YOLOF). This model was pre-trained on [Common Objects in Context (COCO)](https://cocodataset.org/#home) dataset with 80 classes.

Reviewer 2: please provide more detailed description. There are several models in the original repo with different backbones and other interesting features. We need to make clear which model we demonstrate here for customers.

> YOLOF is a simple, fast, and efficient object detector without FPN. Model based on ["You Only Look One-level Feature"](https://arxiv.org/abs/2103.09460) paper. It was implemented in PyTorch\* framework. Model used `CSPDarkNet-53` as backbone. For details see [repository](https://github.com/megvii-model/YOLOF). This model was pre-trained on [Common Objects in Context (COCO)](https://cocodataset.org/#home) dataset with 80 classes.

Reviewer 2: ```suggestion
YOLOF is a simple, fast, and efficient object detector without FPN. Model based on ["You Only Look One-level Feature"](https://arxiv.org/abs/2103.09460) paper. It was implemented in PyTorch\* framework. Model used `DarkNet-53` with Cross Stage Partial blocks as backbone. For details see [repository](https://github.com/megvii-model/YOLOF). This model was pre-trained on [Common Objects in Context (COCO)](https://cocodataset.org/#home) dataset with 80 classes.
```

Reviewer 2: better, but still unclear which model used  from these 3:
<html>
<body>
<!--StartFragment-->

YOLOF_CSP_D_53_DC5_3x | 41.2 | 39
-- | -- | --
YOLOF_CSP_D_53_DC5_9x | 42.8 | 39
YOLOF_CSP_D_53_DC5_9x_stage2_3x | 43.2 | 39

<!--EndFragment-->
</body>
</html>

Author: As far as I understood, this is one architecture, but pre-trained with different hyperparameters (steps, ...). So what info we can demonstrate for customers?

> YOLOF is a simple, fast, and efficient object detector without FPN. Model based on ["You Only Look One-level Feature"](https://arxiv.org/abs/2103.09460) paper. It was implemented in PyTorch\* framework. Model used `DarkNet-53` with Cross Stage Partial blocks as backbone. For details see [repository](https://github.com/megvii-model/YOLOF). This model was pre-trained on [Common Objects in Context (COCO)](https://cocodataset.org/#home) dataset with 80 classes

Reviewer 3: ```suggestion
YOLOF is a simple, fast, and efficient object detector without FPN. Model based on ["You Only Look One-level Feature"](https://arxiv.org/abs/2103.09460) paper. It was implemented in PyTorch\* framework. Model used `DarkNet-53` with Cross Stage Partial blocks as backbone. For details see [repository](https://github.com/megvii-model/YOLOF). This model was pre-trained on [Common Objects in Context (COCO)](https://cocodataset.org/#home) dataset with 80 classes. Mapping of class IDs to label names provided in `<omz_dir>/data/dataset_classes/coco_80cl.txt` file.
```

- `B` - batch size
- `C` - channel
- `H` - height
- `W` - width

Channel order is `BGR`.
Mean values - [103.53, 116.28, 123.675].

### Converted model

Image, name - `image_input`, shape - `1, 3, 608, 608`, format is `B, C, H, W`, where:

- `B` - batch size
- `C` - channel
- `H` - height
- `W` - width

Channel order is `BGR`.

## Output

### Original model

The list of instances. The postprocessing is implemented inside the model and is performed while inference of model. So each instance it is a object with next fields:
- `detection box`
- `label` - predicted class ID
- `score` - onfidence for the predicted class

Detection box has format [`x_min`, `y_min`, `x_max`, `y_max`], where:

- (`x_min`, `y_min`) - coordinates of the top left bounding box corner
- (`x_max`, `y_max`) - coordinates of the bottom right bounding box corner

### Converted model

The array of detection summary info, name - `boxes`, shape - `1, 504, 38, 38`. The anchor values are `16,16,  32,32,  64,64,  128,128,  256,256,  512,512`.

For each case format is `B, N*84, Cx, Cy`, where

- `B` - batch size
- `Cx`, `Cy` - cell index
- `N` - number of detection boxes for cell

> Detection box has format [`x`, `y`, `h`, `w`, `class_no_1`, ..., `class_no_80`], where:

Reviewer 3: ```suggestion
Detection box has format [`x`, `y`, `h`, `w`, `class_id_1`, ..., `class_id_80`], where:
```

- (`x`, `y`) - raw coordinates of box center, multiply by corresponding anchors to get relative to the cell coordinates
- `h`, `w` - raw height and width of box, apply [exponential function](https://en.wikipedia.org/wiki/Exponential_function) and multiply by corresponding anchors to get absolute height and width values

> - `class_no_1`,...,`class_no_80` - probability distribution over the classes in logits format, apply [sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function) to get confidence of each class

Reviewer 3: ```suggestion
- `class_id_1`,...,`class_id_80` - probability distribution over the classes in logits format, apply [sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function) to get confidence of each class
```

> The model was trained on [Common Objects in Context (COCO)](https://cocodataset.org/#home) dataset version with 80 categories of object. Mapping to class names provided in `<omz_dir>/data/dataset_classes/coco_80cl.txt` file.

Reviewer 3: ```suggestion
```
